File: src/lib/data/dataService.ts
import { MongoClient, ObjectId } from 'mongodb';
import { ModelDefinition } from '@/types/modelDefinition';
import { CreateDataRecordInput, DataRecord, ListRecordsQuery, UpdateDataRecordInput } from '@/types/dataRecord';
import { getDataMongoClient, ensureCollection } from '@/lib/db/dataDb';
import { EmbeddingService } from '@/lib/embeddings/embeddingService';

export class DataService {
  private client: MongoClient;
  private model: ModelDefinition;
  private embeddingService: EmbeddingService;

  constructor(model: ModelDefinition) {
    this.client = getDataMongoClient();
    this.model = model;
    this.embeddingService = new EmbeddingService(model);
    // No automatic collection initialization
  }

  // Add manual initialization method
  async initializeCollection(): Promise<void> {
    await ensureCollection(this.model.id);
  }

  private getCollectionName(): string {
    return `data_${this.model.id}`;
  }

  /**
   * Transform database record to client record
   */
  private toClientRecord(dbRecord: any): DataRecord {
    if (!dbRecord) return dbRecord;

    // Extract system fields and other fields
    const { _id, _created_at, _updated_at, _vector, ...fields } = dbRecord;

    // Return client format with flattened structure
    return {
      _id,
      _created_at: new Date(_created_at),
      _updated_at: new Date(_updated_at),
      _vector,
      ...fields
    };
  }

  /**
   * Transform client record to database record
   */
  private toDbRecord(clientRecord: Partial<DataRecord>): any {
    if (!clientRecord) return clientRecord;

    // Extract system fields and other fields
    const { _id, _created_at, _updated_at, _vector, ...fields } = clientRecord;

    // Return flattened database format
    return {
      _id,
      _created_at,
      _updated_at,
      _vector,
      ...fields
    };
  }

  private async validateFields(fields: Record<string, any>): Promise<void> {
    // Validate required fields
    for (const [fieldName, fieldDef] of Object.entries(this.model.fields)) {
      if (fieldDef.required && !(fieldName in fields)) {
        throw new Error(`Missing required field: ${fieldName}`);
      }

      if (fieldName in fields) {
        const value = fields[fieldName];
        
        // Type validation
        switch (fieldDef.type) {
          case 'string':
            if (typeof value !== 'string') {
              throw new Error(`Field ${fieldName} must be a string`);
            }
            break;
          case 'number':
            if (typeof value !== 'number') {
              throw new Error(`Field ${fieldName} must be a number`);
            }
            break;
          case 'boolean':
            if (typeof value !== 'boolean') {
              throw new Error(`Field ${fieldName} must be a boolean`);
            }
            break;
          case 'date':
            if (!(value instanceof Date) && !Date.parse(value)) {
              throw new Error(`Field ${fieldName} must be a valid date`);
            }
            break;
          case 'vector':
            if (!Array.isArray(value) || !value.every(n => typeof n === 'number')) {
              throw new Error(`Field ${fieldName} must be an array of numbers`);
            }
            break;
        }

        // Enum validation
        if (fieldDef.enum && !fieldDef.enum.includes(value)) {
          throw new Error(`Invalid value for field ${fieldName}. Must be one of: ${fieldDef.enum.join(', ')}`);
        }
      }
    }
  }

  async createRecord(input: CreateDataRecordInput): Promise<DataRecord> {
    // Ensure collection exists before creating record
    await this.initializeCollection();
    
    if (!input) {
      throw new Error('Input is required');
    }

    // Filter out any system fields from input
    const { _id, _created_at, _updated_at, _vector, ...fields } = input;

    // Validate fields against model definition
    await this.validateFields(fields);

    const now = new Date();
    const clientRecord: DataRecord = {
      _id: new ObjectId(),
      _created_at: now,
      _updated_at: now,
      ...fields
    };

    let vector: number[] | undefined = undefined;
    // Generate embedding first if enabled
    if (this.model.embedding?.enabled) {
      try {
        vector = await this.embeddingService.updateRecordEmbedding(clientRecord);
      } catch (error: unknown) {
        const message = error instanceof Error ? error.message : 'Unknown error during embedding';
        throw new Error(`Creation failed: ${message}`);
      }
    }

    // Only proceed with database insertion if embedding was successful or not required
    const dbRecord = this.toDbRecord({
      ...clientRecord,
      _vector: vector
    });
    
    const collection = this.client.db().collection(this.getCollectionName());
    await collection.insertOne(dbRecord);

    console.log('[Data] Record created, ID:', clientRecord._id);
    return {
      ...clientRecord,
      _vector: vector
    };
  }

  private excludeVectorData(record: any) {
    const { _vector, ...rest } = record;
    return rest;
  }

  async getRecord(id: string): Promise<DataRecord> {
    const collection = this.client.db().collection(this.getCollectionName());
    const record = await collection.findOne({ _id: new ObjectId(id) });

    if (!record) {
      throw new Error('Record not found');
    }

    return this.excludeVectorData(record);
  }

  async listRecords(query: ListRecordsQuery): Promise<{ records: DataRecord[]; total: number }> {
    const collection = this.client.db().collection(this.getCollectionName());
    const { filter = {}, page = 1, limit = 10 } = query;

    // Transform filters to work with flattened structure
    const finalFilter: Record<string, any> = {};

    // Handle filters directly since fields are at root level
    for (const [key, value] of Object.entries(filter)) {
      if (key === '_id') {
        finalFilter._id = typeof value === 'string' ? new ObjectId(value) : value;
      } else {
        finalFilter[key] = value;
      }
    }

    const [documents, total] = await Promise.all([
      collection
        .find(finalFilter)
        .skip((page - 1) * limit)
        .limit(limit)
        .toArray(),
      collection.countDocuments(finalFilter),
    ]);

    return { 
      records: documents.map(doc => this.excludeVectorData(doc)),
      total 
    };
  }

  async updateRecord(id: string, input: UpdateDataRecordInput): Promise<DataRecord> {
    if (!input) {
      throw new Error('Input is required');
    }

    // Validate fields against model definition
    await this.validateFields(input);

    const collection = this.client.db().collection(this.getCollectionName());
    
    // First get the existing record
    const existingRecord = await collection.findOne({ _id: new ObjectId(id) });
    if (!existingRecord) {
      throw new Error('Record not found');
    }

    // Prepare the updated record
    const updatedRecord = {
      ...existingRecord,
      ...input,
      _updated_at: new Date(),
    };

    const clientRecord = this.toClientRecord(updatedRecord);

    let vector: number[] | undefined = undefined;
    // Update embedding first if enabled
    if (this.model.embedding?.enabled) {
      try {
        vector = await this.embeddingService.updateRecordEmbedding(clientRecord);
      } catch (error: unknown) {
        const message = error instanceof Error ? error.message : 'Unknown error during embedding';
        throw new Error(`Record update failed: ${message}`);
      }
    }

    // Only proceed with database update if embedding was successful or not required
    const finalRecord = {
      ...clientRecord,
      _vector: vector
    };

    const dbRecord = await collection.findOneAndUpdate(
      { _id: new ObjectId(id) },
      { $set: this.toDbRecord(finalRecord) },
      { returnDocument: 'after' }
    );

    if (!dbRecord) {
      throw new Error('Update failed');
    }

    return this.toClientRecord(dbRecord);
  }

  async deleteRecord(id: string): Promise<void> {
    const collection = this.client.db().collection(this.getCollectionName());
    const result = await collection.deleteOne({ _id: new ObjectId(id) });

    if (result.deletedCount === 0) {
      throw new Error('Record not found or delete failed');
    }
  }

  // Add this method to handle partial updates
  async partialUpdateRecord(id: string, input: Partial<UpdateDataRecordInput>): Promise<DataRecord> {
    if (!input) {
      throw new Error('Input is required');
    }

    // Validate fields against model definition
    await this.validateFields(input);

    const collection = this.client.db().collection(this.getCollectionName());
    const dbRecord = await collection.findOneAndUpdate(
      { _id: new ObjectId(id) },
      {
        $set: {
          ...input,
          _updated_at: new Date(),
        },
      },
      { returnDocument: 'after' }
    );

    if (!dbRecord) {
      throw new Error('Record not found or update failed');
    }

    const clientRecord = this.toClientRecord(dbRecord);

    // Update embedding if needed
    if (this.model.embedding?.enabled) {
      try {
        await this.embeddingService.updateRecordEmbedding(clientRecord);
      } catch (error: unknown) {
        await collection.replaceOne({ _id: new ObjectId(id) }, this.toDbRecord(clientRecord));
        const message = error instanceof Error ? error.message : 'Unknown error during embedding';
        throw new Error(`Record update failed: ${message}`);
      }
    }

    return clientRecord;
  }

  // Add this method to handle upserts
  async upsertRecord(id: string | undefined, input: CreateDataRecordInput | UpdateDataRecordInput): Promise<DataRecord> {
    if (id) {
      return this.updateRecord(id, input as UpdateDataRecordInput);
    }
    return this.createRecord(input as CreateDataRecordInput);
  }
} 

File: infinity-sheets/app/services/data_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update, delete, text
from sqlalchemy.sql import func
import uuid
from datetime import datetime
import logging
import json

from app.db.models import ModelData
from app.services.model_service import ModelService
from app.services.embedding_service import EmbeddingService
from app.core.errors import NotFoundError, ValidationError, AuthorizationError

logger = logging.getLogger(__name__)

class DataService:
    def __init__(self, db: AsyncSession):
        self.db = db
        self.model_service = ModelService(db)
        self.embedding_service = EmbeddingService()
    
    async def create_record(self, model_id: uuid.UUID, data: dict, owner_id: uuid.UUID):
        """Create a new data record"""
        
        # Get model definition
        model = await self.model_service.get_model(model_id)
        
        # Validate data against model definition
        await self._validate_data(data, model.fields)
        
        # Generate embedding if enabled
        embedding = None
        if model.embedding and model.embedding.get("enabled", False):
            source_fields = model.embedding.get("source_fields", [])
            if source_fields:
                text_for_embedding = self._get_text_for_embedding(data, source_fields)
                if text_for_embedding:
                    embedding = await self.embedding_service.generate_embedding(text_for_embedding)
        
        # Create record
        record = ModelData(
            model_id=model_id,
            owner_id=owner_id,
            data=data,
            embedding=embedding,
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow()
        )
        
        try:
            # Add to database
            self.db.add(record)
            await self.db.flush()
            await self.db.refresh(record)
            
            logger.info(f"Created data record: {record.id} for model: {model_id}")
            return self._to_client_record(record)
        except Exception as e:
            logger.error(f"Error creating data record: {str(e)}")
            raise ValidationError(f"Failed to create data record: {str(e)}")
    
    async def get_record(self, model_id: uuid.UUID, record_id: uuid.UUID):
        """Get a data record by ID"""
        
        query = select(ModelData).where(
            ModelData.model_id == model_id,
            ModelData.id == record_id
        )
        result = await self.db.execute(query)
        record = result.scalars().first()
        
        if not record:
            logger.warning(f"Record not found: {record_id} for model: {model_id}")
            raise NotFoundError(f"Record with ID {record_id} not found")
        
        return self._to_client_record(record)
    
    async def list_records(self, model_id: uuid.UUID, filter_data: dict = None, page: int = 1, limit: int = 10):
        """List data records with pagination and filtering"""
        # Validate model exists
        await self.model_service.get_model(model_id)
        
        # Build query
        query = select(ModelData).where(ModelData.model_id == model_id)
        
        # Apply filters if provided
        if filter_data:
            for field, value in filter_data.items():
                # For JSON fields, we need to use the -> operator
                if isinstance(value, dict):
                    # Complex filter with operators
                    operator = value.get("operator", "eq")
                    filter_value = value.get("value")
                    
                    if operator == "eq":
                        query = query.where(text(f"data->'{field}' = '{filter_value}'"))
                    elif operator == "neq":
                        query = query.where(text(f"data->'{field}' != '{filter_value}'"))
                    elif operator == "gt":
                        query = query.where(text(f"data->'{field}' > '{filter_value}'"))
                    elif operator == "gte":
                        query = query.where(text(f"data->'{field}' >= '{filter_value}'"))
                    elif operator == "lt":
                        query = query.where(text(f"data->'{field}' < '{filter_value}'"))
                    elif operator == "lte":
                        query = query.where(text(f"data->'{field}' <= '{filter_value}'"))
                    elif operator == "contains":
                        query = query.where(text(f"data->'{field}' LIKE '%{filter_value}%'"))
                else:
                    # Simple equality filter
                    query = query.where(text(f"data->'{field}' = '{value}'"))
        
        # Add pagination
        total_query = select(func.count()).select_from(ModelData).where(ModelData.model_id == model_id)
        total_result = await self.db.execute(total_query)
        total = total_result.scalar()
        
        query = query.order_by(ModelData.created_at.desc())
        query = query.offset((page - 1) * limit).limit(limit)
        
        # Execute query
        result = await self.db.execute(query)
        records = result.scalars().all()
        
        # Convert to client records
        client_records = [self._to_client_record(record) for record in records]
        
        # Calculate pagination metadata
        total_pages = (total + limit - 1) // limit if total > 0 else 1
        
        return {
            "records": client_records,
            "pagination": {
                "page": page,
                "limit": limit,
                "total": total,
                "totalPages": total_pages,
                "hasNextPage": page < total_pages,
                "hasPreviousPage": page > 1
            }
        }
    
    async def update_record(self, model_id: uuid.UUID, record_id: uuid.UUID, data: dict, owner_id: uuid.UUID):
        """Update a data record"""
        
        # Get model definition
        model = await self.model_service.get_model(model_id)
        
        # Get existing record
        query = select(ModelData).where(
            ModelData.model_id == model_id,
            ModelData.id == record_id
        )
        result = await self.db.execute(query)
        record = result.scalars().first()
        
        if not record:
            logger.warning(f"Record not found: {record_id} for model: {model_id}")
            raise NotFoundError(f"Record with ID {record_id} not found")
        
        # Check ownership
        if record.owner_id != owner_id:
            logger.warning(f"Unauthorized record update attempt: {record_id} by {owner_id}")
            raise AuthorizationError("You don't have permission to update this record")
        
        # Validate data against model definition
        await self._validate_data(data, model.fields)
        
        # Generate embedding if enabled
        embedding = record.embedding
        if model.embedding and model.embedding.get("enabled", False):
            source_fields = model.embedding.get("source_fields", [])
            if source_fields:
                text_for_embedding = self._get_text_for_embedding(data, source_fields)
                if text_for_embedding:
                    embedding = await self.embedding_service.generate_embedding(text_for_embedding)
        
        try:
            # Update record
            query = update(ModelData).where(
                ModelData.id == record_id
            ).values(
                data=data,
                embedding=embedding,
                updated_at=datetime.utcnow()
            )
            
            await self.db.execute(query)
            
            # Get updated record
            updated_record = await self.get_record(model_id, record_id)
            logger.info(f"Updated data record: {record_id}")
            return updated_record
        except Exception as e:
            logger.error(f"Error updating data record: {str(e)}")
            raise ValidationError(f"Failed to update data record: {str(e)}")
    
    async def delete_record(self, model_id: uuid.UUID, record_id: uuid.UUID, owner_id: uuid.UUID):
        """Delete a data record"""
        
        # Get existing record
        query = select(ModelData).where(
            ModelData.model_id == model_id,
            ModelData.id == record_id
        )
        result = await self.db.execute(query)
        record = result.scalars().first()
        
        if not record:
            logger.warning(f"Record not found: {record_id} for model: {model_id}")
            raise NotFoundError(f"Record with ID {record_id} not found")
        
        # Check ownership
        if record.owner_id != owner_id:
            logger.warning(f"Unauthorized record deletion attempt: {record_id} by {owner_id}")
            raise AuthorizationError("You don't have permission to delete this record")
        
        try:
            # Delete record
            query = delete(ModelData).where(ModelData.id == record_id)
            await self.db.execute(query)
            
            logger.info(f"Deleted data record: {record_id}")
            return True
        except Exception as e:
            logger.error(f"Error deleting data record: {str(e)}")
            raise ValidationError(f"Failed to delete data record: {str(e)}")
    
    async def search_records(self, model_id: uuid.UUID, query_text: str, filter_data: dict = None, limit: int = 10, min_similarity: float = 0.7):
        """Search for records using vector similarity"""
        # Get model definition
        model = await self.model_service.get_model(model_id)
        
        # Check if embedding is enabled for this model
        if not model.embedding or not model.embedding.get("enabled", False):
            logger.warning(f"Vector search attempted on model without embeddings: {model_id}")
            raise ValidationError("Vector search is not enabled for this model")
        
        # Generate embedding for query text
        query_embedding = await self.embedding_service.generate_embedding(query_text)
        if not query_embedding:
            raise ValidationError("Failed to generate embedding for search query")
        
        # Build query
        # This is a simplified approach - in a production environment, you would use
        # pgvector's vector similarity functions for better performance
        query = select(ModelData).where(ModelData.model_id == model_id)
        
        # Apply filters if provided
        if filter_data:
            for field, value in filter_data.items():
                if isinstance(value, dict):
                    # Complex filter with operators
                    operator = value.get("operator", "eq")
                    filter_value = value.get("value")
                    
                    if operator == "eq":
                        query = query.where(text(f"data->'{field}' = '{filter_value}'"))
                    elif operator == "neq":
                        query = query.where(text(f"data->'{field}' != '{filter_value}'"))
                    elif operator == "gt":
                        query = query.where(text(f"data->'{field}' > '{filter_value}'"))
                    elif operator == "gte":
                        query = query.where(text(f"data->'{field}' >= '{filter_value}'"))
                    elif operator == "lt":
                        query = query.where(text(f"data->'{field}' < '{filter_value}'"))
                    elif operator == "lte":
                        query = query.where(text(f"data->'{field}' <= '{filter_value}'"))
                    elif operator == "contains":
                        query = query.where(text(f"data->'{field}' LIKE '%{filter_value}%'"))
                else:
                    # Simple equality filter
                    query = query.where(text(f"data->'{field}' = '{value}'"))
        
        # Execute query to get all records (with limit)
        # In a production environment, you would use a more efficient approach
        result = await self.db.execute(query.limit(100))  # Limit to avoid processing too many records
        records = result.scalars().all()
        
        # Calculate similarity scores
        search_results = []
        for record in records:
            if record.embedding:
                similarity = self.embedding_service.calculate_similarity(query_embedding, record.embedding)
                if similarity >= min_similarity:
                    search_results.append({
                        "record": self._to_client_record(record),
                        "similarity": similarity
                    })
        
        # Sort by similarity (descending) and apply limit
        search_results.sort(key=lambda x: x["similarity"], reverse=True)
        search_results = search_results[:limit]
        
        return search_results
    
    async def _validate_data(self, data: dict, field_definitions: dict):
        """Validate data against field definitions"""
        if not isinstance(data, dict):
            raise ValidationError("Data must be a dictionary")
        
        # Check required fields
        for field_name, field_def in field_definitions.items():
            if field_def.get("required", False) and field_name not in data:
                raise ValidationError(f"Required field missing: {field_name}")
        
        # Validate field types and constraints
        for field_name, field_value in data.items():
            # Skip if field is not in model definition (could be extra data)
            if field_name not in field_definitions:
                continue
                
            field_def = field_definitions[field_name]
            field_type = field_def.get("type")
            
            # Type validation
            if field_value is not None:
                if field_type == "string" and not isinstance(field_value, str):
                    raise ValidationError(f"Field {field_name} must be a string")
                elif field_type == "number" and not isinstance(field_value, (int, float)):
                    raise ValidationError(f"Field {field_name} must be a number")
                elif field_type == "boolean" and not isinstance(field_value, bool):
                    raise ValidationError(f"Field {field_name} must be a boolean")
                elif field_type == "date":
                    # Date validation is more complex - could be a string in ISO format
                    if not isinstance(field_value, (str, datetime)):
                        raise ValidationError(f"Field {field_name} must be a date string or datetime object")
            
            # Enum validation for string fields
            if field_type == "string" and "enum" in field_def and field_def["enum"]:
                if field_value not in field_def["enum"]:
                    raise ValidationError(f"Field {field_name} must be one of: {', '.join(field_def['enum'])}")
            
            # Range validation for number fields
            if field_type == "number":
                if "min" in field_def and field_value < field_def["min"]:
                    raise ValidationError(f"Field {field_name} must be at least {field_def['min']}")
                if "max" in field_def and field_value > field_def["max"]:
                    raise ValidationError(f"Field {field_name} must be at most {field_def['max']}")
    
    def _get_text_for_embedding(self, data: dict, source_fields: list) -> str:
        """Extract text from specified fields for embedding generation"""
        if not source_fields:
            return ""
            
        text_parts = []
        for field in source_fields:
            if field in data and isinstance(data[field], str):
                text_parts.append(data[field])
        
        return " ".join(text_parts)
    
    def _to_client_record(self, db_record) -> dict:
        """Convert database record to client-friendly format"""
        # Combine metadata and data
        record_data = db_record.data.copy() if db_record.data else {}
        
        # Add metadata with underscore prefix
        record_data["_id"] = str(db_record.id)
        record_data["_created_at"] = db_record.created_at.isoformat() if db_record.created_at else None
        record_data["_updated_at"] = db_record.updated_at.isoformat() if db_record.updated_at else None
        
        # Don't include the embedding vector in the response
        
        return record_data
